From 8cceb96bdba8e50e472f9fcf79d617ae040c7742 Mon Sep 17 00:00:00 2001
From: Ovidiu Panait <ovidiu.panait@windriver.com>
Date: Tue, 2 Nov 2021 09:46:45 +0000
Subject: [PATCH 1/3] Revert "srcu: Use local irq lock instead of irq disable
 regions"

This reverts commit 76f1001e0c23a23145b6a4637144ce115526cc78.

Revert this WR-specific commit in order to backport the more complete upstream
version of this fix:
https://git.kernel.org/pub/scm/linux/kernel/git/rt/linux-stable-rt.git/commit/?h=v4.14-rt&id=b6f2d7e583f19f21d0588f8f74c83dddf877488a

Signed-off-by: Ovidiu Panait <ovidiu.panait@windriver.com>
---
 kernel/rcu/srcutree.c | 8 ++------
 1 file changed, 2 insertions(+), 6 deletions(-)

diff --git a/kernel/rcu/srcutree.c b/kernel/rcu/srcutree.c
index b94748a181fe..876d9781c8be 100644
--- a/kernel/rcu/srcutree.c
+++ b/kernel/rcu/srcutree.c
@@ -36,7 +36,6 @@
 #include <linux/delay.h>
 #include <linux/module.h>
 #include <linux/srcu.h>
-#include <linux/locallock.h>
 
 #include "rcu.h"
 #include "rcu_segcblist.h"
@@ -773,8 +772,6 @@ static bool srcu_might_be_idle(struct srcu_struct *sp)
  * srcu_read_lock(), and srcu_read_unlock() that are all passed the same
  * srcu_struct structure.
  */
-static DEFINE_LOCAL_IRQ_LOCK(callsrcu_lock);
-
 void __call_srcu(struct srcu_struct *sp, struct rcu_head *rhp,
 		 rcu_callback_t func, bool do_norm)
 {
@@ -786,7 +783,7 @@ void __call_srcu(struct srcu_struct *sp, struct rcu_head *rhp,
 
 	check_init_srcu_struct(sp);
 	rhp->func = func;
-	local_lock_irqsave(callsrcu_lock, flags);
+	local_irq_save(flags);
 	sdp = this_cpu_ptr(sp->sda);
 	spin_lock(&sdp->lock);
 	rcu_segcblist_enqueue(&sdp->srcu_cblist, rhp, false);
@@ -802,8 +799,7 @@ void __call_srcu(struct srcu_struct *sp, struct rcu_head *rhp,
 		sdp->srcu_gp_seq_needed_exp = s;
 		needexp = true;
 	}
-	spin_unlock(&sdp->lock);
-	local_unlock_irqrestore(callsrcu_lock, flags);
+	spin_unlock_irqrestore(&sdp->lock, flags);
 	if (needgp)
 		srcu_funnel_gp_start(sp, sdp, s, do_norm);
 	else if (needexp)
-- 
2.31.1

